### PR TITLE
Replace TemporaryFile() with io.BytesIO()

### PR DESCRIPTION
**Pull request by Ankit-Dhankhar**
Using similar syntax as torch_tensor_serializer/desearializer. #2357

### CODE REVIEW

**Comment by midokura-silvia** [comment]
Ok, I just confirmed that there is a unit test. Could we improve the unit test by using `@pytest.mark.parametrize` and call it for let's say 4 different arrays? Some combination of float and int arrays, with non-symmetric dimensions? The current choice of `numpy.ones((10, 10))` feels a bit limited.

**Comment by Ankit-Dhankhar** [comment]
Sure I will add test cases.

**Comment by Ankit-Dhankhar** [comment]
I have added test cases for random integer arrays of non-symmetric dimensions.
But for array with float value it is giving `assert tensor_serialized[0] != serde.NO_COMPRESSION` this false.

File: syft/serde/torch_serde.py [hunk 1]
```diff
@@ -59,30 +58,38 @@ def _deserialize_tensor(tensor_bin) -> torch.Tensor:
...
+    Returns
+        A serialized version of the input tensor
     """
     if tensor.requires_grad:
         warnings.warn(
             "Torch to Numpy serializer can only be used with tensors that do not require grad. "
             "Detaching tensor to continue"
         )
-        tensor = tensor.detach()
+        tensor = torch.detach()
```
[the remark this answers is not in the document]
Comment (Ankit-Dhankhar) [reply syft/serde/torch_serde.py]: It is a typo mistake. As there was no test case for "requires_grad=True", I missed it.

File: test/test_serde.py [hunk 2]
```diff
@@ -568,14 +568,20 @@ def test_pointer_tensor_detail(id):
     assert (x_back == 2 * x).all()
 
 
-def test_numpy_tensor_serde():
+@pytest.mark.parametrize(
+    "tensor",
+    [
+        (torch.tensor(numpy.ones((10, 10)), requires_grad=False)),
```
[the remark this answers is not in the document]
Comment (Ankit-Dhankhar) [reply test/test_serde.py]: Yes, I also believe that we need it else mistake like pointed in the previous comment will not be detected.

File: syft/serde/torch_serde.py [hunk 3]
```diff
@@ -59,30 +58,38 @@ def _deserialize_tensor(tensor_bin) -> torch.Tensor:
 
 def numpy_tensor_serializer(tensor: torch.Tensor) -> bin:
     """Strategy to serialize a tensor using numpy npy format.
-    If tensor requires to calculate gradients, it will be detached.
+    If tensor requires to calculate gradients, it will detached.
```
[the remark this answers is not in the document]
Comment (Ankit-Dhankhar) [reply syft/serde/torch_serde.py]: Updated

**Comment by Ankit-Dhankhar** [comment]
I have added test-cases with requires_grad=True but tensor generated with torch.randn([dim]) fails this. Is this expected?

> I have added test cases for random integer arrays of non-symmetric dimensions.
> But for array with float value it is giving `assert tensor_serialized[0] != serde.NO_COMPRESSION` this false.

File: syft/serde/torch_serde.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ankit-Dhankhar) [reply syft/serde/torch_serde.py]: Updated

File: test/test_serde.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ankit-Dhankhar) [reply test/test_serde.py]: Added test case

**Comment by midokura-silvia** [comment]
@robert-wagner any update from your side?

**Comment by iamtrask** [comment]
@midokura-silvia - if it's been ~7 days or more after checking in with someone who reviewed - and you feel the PR should be approved - you may go ahead and merge.

**Comment by iamtrask** [comment]
(in this case it's been 3 months - more than enough time)

**Comment by gmuraru** [comment]
Can I pick this up?

**Comment by karlhigley** [comment]
@Ankit-Dhankhar I cloned your fork, squashed the commits in this PR, and rebased it on current PySyft master. You can see the [resulting diff](https://github.com/OpenMined/PySyft/commit/39b7f23a25834299f99d6bee062efee0002499d2) over here on the `pr/2376` branch.

Sorry it's taken so long! Okay to force-push the result to your branch/PR so we can finally get this merged? 😄
